lldb/source/Core/DILAST.cpp

```diff
@@ -0,0 +1,560 @@
+//===-- DILAST.cpp --------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "lldb/Core/DILAST.h"
+#include "lldb/API/SBType.h"
+#include "lldb/Core/ValueObjectRegister.h"
+#include "lldb/Core/ValueObjectVariable.h"
+#include "lldb/Symbol/TypeList.h"
+#include "lldb/Symbol/VariableList.h"
+#include "lldb/Target/LanguageRuntime.h"
+#include "lldb/Target/RegisterContext.h"
+#include "llvm/ADT/StringRef.h"
+
+#include <vector>
+
+namespace lldb_private {
+
+namespace DIL {
+
+lldb::ValueObjectSP
+GetDynamicOrSyntheticValue(lldb::ValueObjectSP in_valobj_sp,
+                           lldb::DynamicValueType use_dynamic,
+                           bool use_synthetic) {
+  Status error;
+
+  if (!in_valobj_sp) {
+    error.SetErrorString("invalid value object");
+    return in_valobj_sp;
+  }
+
+  lldb::ValueObjectSP value_sp = in_valobj_sp;
+
+  Target *target = value_sp->GetTargetSP().get();
+  // If this ValueObject holds an error, then it is valuable for that.
+  if (value_sp->GetError().Fail())
+    return value_sp;
+
+  if (!target)
+    return lldb::ValueObjectSP();
+
+  if (use_dynamic != lldb::eNoDynamicValues) {
+    lldb::ValueObjectSP dynamic_sp = value_sp->GetDynamicValue(use_dynamic);
+    if (dynamic_sp)
+      value_sp = dynamic_sp;
+  }
+
+  if (use_synthetic) {
+    lldb::ValueObjectSP synthetic_sp = value_sp->GetSyntheticValue();
+    if (synthetic_sp)
+      value_sp = synthetic_sp;
+  }
+
+  if (!value_sp)
+    error.SetErrorString("invalid value object");
+
+  return value_sp;
+}
+
+CompilerType DILASTNode::GetDereferencedResultType() const {
+  auto type = result_type();
+  return type.IsReferenceType() ? type.GetNonReferenceType() : type;
+}
+
+std::optional<MemberInfo>
```

Comment:
I've tried that, but the lhs_val_sp does not always have the right type for finding the field name. :-(. One of the simplest examples: in one test case I have a class "C".  I declare an array variable "C c_arr[2];".  The ValueObject for 'c_arr' has the type "C[2]", rather than "C".  Type "C" has named field members, but type "C[2]" does not, so GetChildMemberWithName doesn't work on it.  I have *many* different test cases where it turns out the the ValueObject associated with the variable doesn't have quite the right type for using GetChildMemberWithName. By the time I've special-cased and updated the code to handle all of them, it doesn't look any better than the function I have here.

Reply:
Why do we actually want this function to work on something like `C[2]`? That's definitely not what I would have expected from the function signature (nor from its implementation, actually).

Even that was the  behavior we want (which I really hope isn't the case), the other implementation might still make more sense, because it would give you a chance to explain what you're doing and why. Right now, this function looks like it's trying to reimplement `TypeSystemClang::GetIndexOfChildMemberWithName` without any indication of why. I wouldn't be surprised if it didn't work with other languages due to its reliance on how TypeSystemClang or C++ represent certain constructs.

(This extraction of information from the types could be part of why you're needing to detect if something is a smart pointer, etc. We currently don't have a way to tell if a type represents a smart pointer -- we can only do that for a specific value of a that type)

Reply:
This function is called from the parser when it is trying to parse what ought to be a field/member name (i.e. it's gotten through <stuff-before-field-reference> <dot-or-arrow> field_name).  The "stuff-before-field-reference" can be a simple variable name, or something more complex; in this case it was an array subscript, 'c_arr[0]'. (The full command was 'frame variable c_arr[0].field_').  This function determines if the 'stuff-before-field-reference' has a valid field/member with the given name, and the index of that child/member (or set of potential valid indices) and returns that in the "idx" parameter, along with the full information about the field, which gets returned in the MemberInfo struct (the main return value).  This information is later used by the evaluator (once we have finished parsing the entire input) to actually find and return the value the user was asking for.

Since  whether or not the field/member name is valid really depends on the type, I do not understand why you seem to feel it is wrong to use the type (the second parameter) of the "stuff-before-field-reference" to check to see if it is valid and (if it's valid) what its position is?